File: archive/python-src/output/trace_map.py
#!/usr/bin/env python3
"""Generate trace-map.json from extraction and matching results.

This is the universal output format consumed by:
- LSP server (editor visualization)
- HTML viewer (if built later)
- Any downstream tool

The trace-map is the single source of truth for all visualization.
"""

import json
import sys
from datetime import datetime, timezone
from pathlib import Path

TRACE_MAP_VERSION = 1


def build_trace_map(
    traces: list[dict],
    source_files: list[str],
    target_files: list[str],
    untraced_clauses: list[dict] | None = None,
    metadata: dict | None = None,
    reverse_traces: list[dict] | None = None,
) -> dict:
    """Build a trace-map.json structure.

    Args:
        traces: list of trace dicts, each with:
            - id, text, type, source (file/line/col), status, refs, topics, etc.
        source_files: list of source file paths
        target_files: list of target file paths
        untraced_clauses: clauses not consumed by any trace (from remainder tracker)
        metadata: optional extra metadata
        reverse_traces: target-extracted claims with reverse traceability status

    Returns:
        Complete trace-map dict ready for JSON serialization.
    """
    summary = compute_summary(traces, reverse_traces)

    trace_map = {
        'version': TRACE_MAP_VERSION,
        'generated': datetime.now(timezone.utc).isoformat(),
        'source_files': source_files,
        'target_files': target_files,
        'traces': traces,
        'untraced_clauses': [
            {'clause_id': c.get('clause_id', ''), 'text': c.get('text', '')}
            for c in (untraced_clauses or [])
        ],
        'summary': summary,
    }

    if reverse_traces is not None:
        trace_map['reverse_traces'] = reverse_traces

    if metadata:
        trace_map['metadata'] = metadata

    return trace_map


def compute_summary(traces: list[dict], reverse_traces: list[dict] | None = None) -> dict:
    """Compute coverage summary from traces list, optionally including reverse stats."""
    status_counts = {}
    for trace in traces:
        status = trace.get('status', 'UNKNOWN')
        status_counts[status] = status_counts.get(status, 0) + 1

    total = len(traces)
    covered = status_counts.get('COVERED', 0)
    partial = status_counts.get('PARTIAL', 0)
    missing = status_counts.get('MISSING', 0)

    checkable = covered + partial + missing
    score = (covered + partial * 0.5) / checkable * 100 if checkable > 0 else 0

    summary = {
        'total': total,
        'covered': covered,
        'partial': partial,
        'missing': missing,
        'deferred': status_counts.get('DEFERRED', 0),
        'superseded': status_counts.get('SUPERSEDED', 0),
        'na': status_counts.get('N/A', 0),
        'coverage_score_pct': round(score, 1),
    }

    if reverse_traces is not None:
        rev_counts = {}
        for rt in reverse_traces:
            st = rt.get('status', 'UNKNOWN')
            rev_counts[st] = rev_counts.get(st, 0) + 1

        rev_traced = rev_counts.get('TRACED', 0)
        rev_partial = rev_counts.get('PARTIAL_SOURCE', 0)
        rev_untraced = rev_counts.get('UNTRACED_IN_SOURCE', 0)
        rev_checkable = rev_traced + rev_partial + rev_untraced

        rev_score = (
            (rev_traced + rev_partial * 0.5) / rev_checkable * 100
            if rev_checkable > 0 else 0
        )

        summary['reverse_total'] = len(reverse_traces)
        summary['reverse_traced'] = rev_traced
        summary['reverse_partial_source'] = rev_partial
        summary['reverse_untraced'] = rev_untraced
        summary['reverse_coverage_pct'] = round(rev_score, 1)

    return summary


def build_source_annotations(traces: list[dict]) -> dict:
    """Build per-file annotation map for source documents.

    Returns: {filepath: [{range, status, refs, trace_id}]}
    Used by LSP server to publish diagnostics on source files.
    """
    annotations = {}

    for trace in traces:
        source = trace.get('source', {})
        filepath = source.get('file')
        if not filepath:
            continue

        if filepath not in annotations:
            annotations[filepath] = []

        annotations[filepath].append({
            'range': {
                'start': {'line': source.get('line', 0), 'col': source.get('col_start', 0)},
                'end': {'line': source.get('line', 0), 'col': source.get('col_end', 0)},
            },
            'status': trace.get('status', 'UNKNOWN'),
            'trace_id': trace.get('id'),
            'trace_text': trace.get('text', ''),
            'refs': trace.get('refs', []),
        })

    return annotations


def build_target_annotations(traces: list[dict], reverse_traces: list[dict] | None = None) -> dict:
    """Build per-file annotation map for target documents.

    Returns: {filepath: [{range, trace_ids, trace_texts}]}
    Used by LSP server to publish diagnostics and references on target files.
    Includes UNTRACED_IN_SOURCE diagnostics from reverse traces.
    """
    annotations = {}

    for trace in traces:
        for ref in trace.get('refs', []):
            filepath = ref.get('file')
            if not filepath:
                continue

            if filepath not in annotations:
                annotations[filepath] = []

            annotations[filepath].append({
                'line': ref.get('line', 0),
                'section': ref.get('section', ''),
                'trace_id': trace.get('id'),
                'trace_text': trace.get('text', ''),
                'status': trace.get('status', 'UNKNOWN'),
                'source': trace.get('source', {}),
            })

    for rt in (reverse_traces or []):
        if rt.get('status') not in ('UNTRACED_IN_SOURCE', 'PARTIAL_SOURCE'):
            continue

        source = rt.get('source', {})
        filepath = source.get('file')
        if not filepath:
            continue

        if filepath not in annotations:
            annotations[filepath] = []

        annotations[filepath].append({
            'line': source.get('line', 0),
            'section': '',
            'trace_id': rt.get('id'),
            'trace_text': rt.get('text', ''),
            'status': rt.get('status'),
            'nearest_source_trace': rt.get('nearest_source_trace'),
            'similarity_note': rt.get('similarity_note', ''),
        })

    return annotations


def main():
    """CLI: build trace-map from traces JSON."""
    if len(sys.argv) < 2:
        print("Usage: trace_map.py <traces.json> [output.json]", file=sys.stderr)
        print("  traces.json must have: traces, source_files, target_files", file=sys.stderr)
        print("  Optional fields: untraced_clauses, metadata", file=sys.stderr)
        sys.exit(1)

    data = json.loads(Path(sys.argv[1]).read_text(encoding='utf-8'))

    trace_map = build_trace_map(
        traces=data.get('traces', []),
        source_files=data.get('source_files', []),
        target_files=data.get('target_files', []),
        untraced_clauses=data.get('untraced_clauses', []),
        metadata=data.get('metadata'),
        reverse_traces=data.get('reverse_traces'),
    )

    output_path = Path(sys.argv[2]) if len(sys.argv) > 2 else None

    if output_path:
        output_path.write_text(
            json.dumps(trace_map, ensure_ascii=False, indent=2),
            encoding='utf-8',
        )
        print(f"Written to {output_path}", file=sys.stderr)
    else:
        json.dump(trace_map, sys.stdout, ensure_ascii=False, indent=2)


if __name__ == '__main__':
    main()

File: src/output/trace-map.ts
export const TRACE_MAP_VERSION = 1;

interface TraceSource {
  file?: string;
  line?: number;
  col_start?: number;
  col_end?: number;
}

interface TraceRef {
  file?: string;
  section?: string;
  line?: number;
}

interface TraceInput {
  id?: string;
  text?: string;
  type?: string;
  source?: TraceSource;
  status?: string;
  refs?: TraceRef[];
  topics?: string[];
  nearest_source_trace?: string;
  similarity_note?: string;
  [key: string]: unknown;
}

interface UntracedClause {
  clause_id: string;
  text: string;
}

interface AnnotationRange {
  start: { line: number; col: number };
  end: { line: number; col: number };
}

export interface SourceAnnotation {
  range: AnnotationRange;
  status: string;
  trace_id: string | undefined;
  trace_text: string;
  refs: unknown[];
}

export interface TargetAnnotation {
  line: number;
  section: string;
  trace_id: string | undefined;
  trace_text: string;
  status: string;
  source?: TraceSource | Record<string, unknown>;
  nearest_source_trace?: string;
  similarity_note?: string;
}

export interface Summary {
  total: number;
  covered: number;
  partial: number;
  missing: number;
  deferred: number;
  superseded: number;
  na: number;
  coverage_score_pct: number;
  reverse_total?: number;
  reverse_traced?: number;
  reverse_partial_source?: number;
  reverse_untraced?: number;
  reverse_coverage_pct?: number;
}

export interface TraceMap {
  version: number;
  generated: string;
  source_files: string[];
  target_files: string[];
  traces: TraceInput[];
  untraced_clauses: UntracedClause[];
  summary: Summary;
  reverse_traces?: TraceInput[];
  metadata?: Record<string, unknown>;
}

export function buildTraceMap(
  traces: TraceInput[],
  sourceFiles: string[],
  targetFiles: string[],
  untracedClauses: Array<{ clause_id?: string; text?: string }> | null = null,
  metadata: Record<string, unknown> | null = null,
  reverseTraces: TraceInput[] | null = null,
): TraceMap {
  const summary = computeSummary(traces, reverseTraces);

  const traceMap: TraceMap = {
    version: TRACE_MAP_VERSION,
    generated: new Date().toISOString(),
    source_files: sourceFiles,
    target_files: targetFiles,
    traces,
    untraced_clauses: (untracedClauses ?? []).map(c => ({
      clause_id: c.clause_id ?? '',
      text: c.text ?? '',
    })),
    summary,
  };

  if (reverseTraces !== null) {
    traceMap.reverse_traces = reverseTraces;
  }

  if (metadata) {
    traceMap.metadata = metadata;
  }

  return traceMap;
}

export function computeSummary(
  traces: TraceInput[],
  reverseTraces: TraceInput[] | null = null,
): Summary {
  const statusCounts: Record<string, number> = {};
  for (const trace of traces) {
    const status = trace.status ?? 'UNKNOWN';
    statusCounts[status] = (statusCounts[status] ?? 0) + 1;
  }

  const total = traces.length;
  const covered = statusCounts['COVERED'] ?? 0;
  const partial = statusCounts['PARTIAL'] ?? 0;
  const missing = statusCounts['MISSING'] ?? 0;

  const checkable = covered + partial + missing;
  const score = checkable > 0
    ? (covered + partial * 0.5) / checkable * 100
    : 0;

  const summary: Summary = {
    total,
    covered,
    partial,
    missing,
    deferred: statusCounts['DEFERRED'] ?? 0,
    superseded: statusCounts['SUPERSEDED'] ?? 0,
    na: statusCounts['N/A'] ?? 0,
    coverage_score_pct: Math.round(score * 10) / 10,
  };

  if (reverseTraces !== null) {
    const revCounts: Record<string, number> = {};
    for (const rt of reverseTraces) {
      const st = rt.status ?? 'UNKNOWN';
      revCounts[st] = (revCounts[st] ?? 0) + 1;
    }

    const revTraced = revCounts['TRACED'] ?? 0;
    const revPartial = revCounts['PARTIAL_SOURCE'] ?? 0;
    const revUntraced = revCounts['UNTRACED_IN_SOURCE'] ?? 0;
    const revCheckable = revTraced + revPartial + revUntraced;

    const revScore = revCheckable > 0
      ? (revTraced + revPartial * 0.5) / revCheckable * 100
      : 0;

    summary.reverse_total = reverseTraces.length;
    summary.reverse_traced = revTraced;
    summary.reverse_partial_source = revPartial;
    summary.reverse_untraced = revUntraced;
    summary.reverse_coverage_pct = Math.round(revScore * 10) / 10;
  }

  return summary;
}

export function buildSourceAnnotations(traces: TraceInput[]): Record<string, SourceAnnotation[]> {
  const annotations: Record<string, SourceAnnotation[]> = {};

  for (const trace of traces) {
    const source = trace.source;
    const filepath = source?.file;
    if (!filepath) continue;

    if (!annotations[filepath]) {
      annotations[filepath] = [];
    }

    annotations[filepath].push({
      range: {
        start: { line: source.line ?? 0, col: source.col_start ?? 0 },
        end: { line: source.line ?? 0, col: source.col_end ?? 0 },
      },
      status: trace.status ?? 'UNKNOWN',
      trace_id: trace.id,
      trace_text: trace.text ?? '',
      refs: trace.refs ?? [],
    });
  }

  return annotations;
}

export function buildTargetAnnotations(
  traces: TraceInput[],
  reverseTraces: TraceInput[] | null = null,
): Record<string, TargetAnnotation[]> {
  const annotations: Record<string, TargetAnnotation[]> = {};

  for (const trace of traces) {
    for (const ref of (trace.refs ?? [])) {
      const filepath = ref.file;
      if (!filepath) continue;

      if (!annotations[filepath]) {
        annotations[filepath] = [];
      }

      annotations[filepath].push({
        line: ref.line ?? 0,
        section: ref.section ?? '',
        trace_id: trace.id,
        trace_text: trace.text ?? '',
        status: trace.status ?? 'UNKNOWN',
        source: trace.source ?? {},
      });
    }
  }

  for (const rt of (reverseTraces ?? [])) {
    if (rt.status !== 'UNTRACED_IN_SOURCE' && rt.status !== 'PARTIAL_SOURCE') {
      continue;
    }

    const source = rt.source;
    const filepath = source?.file;
    if (!filepath) continue;

    if (!annotations[filepath]) {
      annotations[filepath] = [];
    }

    annotations[filepath].push({
      line: source.line ?? 0,
      section: '',
      trace_id: rt.id,
      trace_text: rt.text ?? '',
      status: rt.status!,
      nearest_source_trace: rt.nearest_source_trace,
      similarity_note: rt.similarity_note ?? '',
    });
  }

  return annotations;
}
